pkg/algorithm/caspool/v1alpha1/auto.go

```diff
@@ -0,0 +1,282 @@
+	err := op.validateAuto()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "validation for auto cspc failed")
+	}
+
+	// Get the eligible node and disk associated with this node for  pool creation(rather instead of pool creation,
+	// call it pool operation as actions other than creation can also be possible).
+	nodeName, diskList, err := op.getEligibleDisks()
+
+	// With the available disk list, for a disk group
+	diskGroups := op.getDiskGroups(diskList)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get usable node for spc %s", op.CspcObject.Object.Name)
+	}
+
+	// get the device ID for the disk list
+	// NOTE: Actually, for disk, we have dev link in case of physical disk and dev path for sparse disk
+	// We try to get dev link first and if it is not present we get dev path.
+	// Just for naming convention we call it device ID (or disk device ID)which can imply to either dev link or dev path.
+	diskDeviceIDMap, err := op.getDiskDeviceIDMapForDiskAPIList(diskList)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not form disk device ID map for %s", op.CspcObject.Object.Name)
+	}
+
+	spcObject := op.CspcObject.Object
+	cp := caspool.NewBuilder().
+		WithDiskType(spcObject.Spec.Type).
+		WithPoolType(op.CspcObject.Object.Spec.PoolSpec.PoolType).
+		WithAnnotations(op.CspcObject.GetAnnotations()).
+		WithDiskGroup(diskGroups).
+		WithCasTemplateName(op.CspcObject.GetCASTName()).
+		WithCspcName(op.CspcObject.Object.Name).
+		WithNodeName(nodeName).
+		WithDiskDeviceIDMap(diskDeviceIDMap).
+		Build().Object
+	return cp, nil
+}
+
+// validateAuto validates auto CSPC
+func (op *Operations) validateAuto() error {
+	err := NewOperationsBuilderForObject(op).
+		WithCheckf(IsMaxPoolNotNil(), "max pool is nil").
+		WithCheckf(IsPoolTypeValid(), "pool type is not valid").
+		WithCheckf(IsTypeValid(), "disk type is not valid").
+		Validate()
+	if err != nil {
+		return errors.Wrapf(err, "validation for cstorpoolcluster %s failed", op.CspcObject.Object.Name)
+	}
+	return nil
+}
+
+// getDisks returns a list of disk
+// The returned disk is
+// 1. Active
+// 2. Not used in other cstor-pool
+// 3. is not attached to a node where a cstor-pool has already been created for the given CSPC.
+func (op *Operations) getDisks() (*apisv1alpha1.DiskList, error) {
+	diskAPIList, err := op.DiskClient.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list disk")
+	}
+
+	usedDiskMap, err := op.getUsedDiskMap()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get used disk map")
+	}
+
+	usedNodeMap, err := op.getUsedNode()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get used node map")
+	}
+
+	newDiskAPIList := disk.NewListBuilderForAPIList(diskAPIList).DiskList.
+		Filter(disk.IsActive(), disk.IsType(op.CspcObject.Object.Spec.Type), disk.IsUsable(usedDiskMap), disk.IsUsableNode(usedNodeMap)).
+		ObjectList
+
+	if len(newDiskAPIList.Items) == 0 {
```

Comment:
IMO this should be dealt by the caller code.
This function/method's job is to get a list of free disks.
It might return 0 if there are no free disks.

Reply:
Alternatively, we can have a new method such as `getDisksFreeMust()` that in turn calls `getDisksFree()`